### DISCUSSION TITLE
Nuxt Assets

### DISCUSSION THREAD

**Discussion by danielroe** [nuxt/nuxt#22012] [RFCs]
# Nuxt Assets RFC

## Summary

Often users need to reference third-party assets (like fonts or scripts) that are effectively dependencies of their built application. These third-party dependencies can often contribute to negative performance, whether because they are hosted with different caching strategies, because they are render-blocking when loaded, or simply because the browser needs to perform a request to a different origin.

The `nuxt/assets` integration is a new addition to the Nuxt framework that allows developers and modules to handle loading third-party assets.

Nuxt will ship with support for different types of assets to be handled out-of-the-box, but modules can also register their own strategies for handling assets.

## Implementation

Nuxt will expose a new top-level configuration option: `assets`, with a number of properties. An ad-hoc module, `nuxt/assets`, will be automatically enabled and will handle the implementation and configuration.

### Strategies

By default, the following strategies will be available:

#### `public`

This is a build-time strategy that will download resources and include them in your app's static files. Assets will be cached between builds to save on network bandwidth.

**Inspirations**: https://github.com/antfu/vite-plugin-remote-assets and https://github.com/nuxt-modules/google-fonts

#### `bundle`

This is a build-time strategy that will transparently download resources and include them in the build. This will enable tree-shaking and other transforms of the remote resources. Assets will be cached between builds to save on network bandwidth.

#### `proxy`

This is a runtime strategy that will transparently proxy access to resources through the runtime Nitro server. (It is not compatible with a static website in most cases, though on some static providers, like Netlify and Vercel, this will be possible as well.)

#### Custom strategies

You can register your own strategies too.

```ts
export default defineNuxtConfig({
  assets: {
    strategies: {
      partytown: {
        setup (nuxt) {
          // Setup if required, such as injecting vite/webpack plugins
          // or setting up an environment
        },
        async resolveId (id, context) {
          // You can do anything at this point, but you must return a string
          // or a callback 'task' function that will be run in parallel with other
          // tasks. It is necessary for any code that can be executed lazily without
          // blocking the build process.
        },
      }
    }
  }
})
```

### API

You (or modules you use) can register rules to enable asset strategies for particular resources. By default Nuxt provides the following shortcuts to configure default behaviour, though this can be customised.

```ts
export default defineNuxtConfig({
  assets: {
    rules: {
      font: {
        strategy: 'public',
        test: /\.(woff2?|eot|ttf|otf)$|design-kit/,
        shouldPreload: false,
        shouldPrefetch: false
      },
      script: {
        strategy: 'bundle',
        test: /\.(js|jsx|ts|tsx)$/
      },
      css: {
        strategy: 'bundle',
        test: /\.(css|less|sass|scss|styl)$/
      }
    }
  }
})
```

**Comment by franklin-tina** [comment]
These are great!

**Comment by Barbapapazes** [comment]
How will this works with nuxt/fonts and nuxt/scripts?

**Reply by danielroe** [reply]
By default, remote fonts (like google fonts) will use `public` asset strategy to ensure they are downloaded and part of the build. With scripts, it will be possible to load scripts via partytown, which might use the `proxy` strategy to ensure there are no CORS issues.

**Reply by PassiDel** [reply]
Thats a great was to use Google Fonts GDPR-compliant, which is especially important for Germany!🎉

**Comment by Aareksio** [comment]
1. How does using the asset look? Will all uses be automagically replaced with local URLs? Eg:

```css
@font-face {
  src: local("Lato"), url("https://external.cdn.com/fonts/Lato.ttf");
}
```

2. Will it be possible to use (`bundle`) assets in JS? I see configuration for `test` only, so assuming no dynamic imports.

```ts
import logoNuxt from 'https://external.cdn.com/images/nuxt.svg'
const badge = { label: 'nuxt', image: logoNuxt }
```

3. How would it be possible to import external asset in a manner allowing for tree-shaking?

> This will enable tree-shaking and other transforms of the remote resources.

**Reply by danielroe** [reply]
1. Exactly.
2. It should work just as though they were local assets. 👍
3. I hope to have an example ready shortly to show you, but basically it integrates with the bundler so as far as Vite is concerned, it's a local asset.

**Comment by flozero** [comment]
Hum so basically you create modules that are intern with global config ? or will this be different from a module ?
